### PR TITLE
Add opencode configuration: skills, kernel-dev agent, and project instructions

### PR DESCRIPTION
**Pull request by Khadyatillo1902**
### Files added
**opencode configuration:**
- `.opencode/opencode.json` — Project config with model, bash permissions, and agent registry
- `.opencode/agent/kernel-dev.md` — GPU kernel development subagent implementing the MANTIS optimization loop (Measure→Analyze→Nominate→Triage→Implement→Summarize)
- `.opencode/command/kernel-opt.md` — `/kernel-opt` command for automated kernel optimization workflow
- `.opencode/skills/gpu-kernel-dev/SKILL.md` — General GPU optimization patterns, memory hierarchy, 5-stage correctness harness (from AutoKernel)
- `.opencode/skills/cuda-kernel/SKILL.md` — CUDA JIT compilation, WMMA tensor cores, TMA, bank conflict avoidance
- `.opencode/skills/triton-kernel/SKILL.md` — Triton autotuning, epilogue fusion, DRTriton/CSP-DAG patterns
- `.opencode/skills/metal-kernel/SKILL.md` — Metal Shading Language, SIMD groups, 1-bit GEMM, StreamK scheduling (from PR #609)

**Project instructions:**
- `AGENTS.md` — Repository layout, JIT compilation conventions, platform detection, kernel coding rules

**Docker sbx sandbox:**
- `sbx-kits/flash-kernel-kit/spec.yaml` — Declarative kit that auto-installs PyTorch + CUDA 12.4 + Triton 3.4 + SGLang into isolated microVM. Validated with `sbx kit validate`.
- `sandbox/run.sh` — Quick launcher: `./sandbox/run.sh` (opencode), `./sandbox/run.sh agent topk` (kernel agent), `./sandbox/run.sh shell` (enter sandbox)

**Setup guide:**
- `SETUP.md` — Getting started: install sbx, launch opencode in sandbox, manage sandboxes, cheat sheet

### How to test

```bash
# Install sbx
curl -fsSL https://get.docker.com | sudo REPO_ONLY=1 sh
sudo apt-get install -y docker-sbx
sbx login

# Validate the kit
sbx kit validate sbx-kits/flash-kernel-kit/

# Launch
sbx run opencode --kit ./sbx-kits/flash-kernel-kit/
# Then type: /kernel-opt
```

### Paper references

- **AutoKernel**: 5-stage safety harness (smoke, shape sweep, stability, determinism, edge cases) + Amdahl's Law prioritization
- **DRTriton**: CSP-DAG synthetic data generation, curriculum RL with decoupled rewards, test-time kernel search
- **SOLAR**: Speed-of-Light (SOL) analysis, MANTIS optimization loop, integrity checking (anti-gaming)

### CODE REVIEW

**Comment by yiakwy-xpu-ml-framework-team** [comment]
Thanks for adding agent skills (prompt) to the project.

I will review the content soon.

File: .opencode/skills/cuda-kernel/SKILL.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        sources=[str(KERNEL_PATH / "csrc" / "my_kernel.cu")],
+        extra_cflags=["-O2"],
+        extra_cuda_cflags=["-O2", "-arch=sm_90", "-DENABLE_HOPPER=1"],
+        verbose=True,
+    )
+```
+
+### Platform Flags
+- Hopper (SM 90): `-DENABLE_HOPPER=1 -arch=sm_90`
+- Ampere (SM 80): `-arch=sm_80`
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/cuda-kernel/SKILL.md:25]: @Khadyatillo1902 This prompt should be removed

Since we won't support it

File: .opencode/skills/cuda-kernel/SKILL.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+## Tensor Core Programming (Hopper WGMMA)
+
+### Warp Matrix Multiply-Accumulate (WMMA)
+```cpp
+// 16x16x16 half-precision
+wmma::fragment<wmma::matrix_a, 16, 16, 16, half, wmma::row_major> a_frag;
+wmma::fragment<wmma::matrix_b, 16, 16, 16, half, wmma::col_major> b_frag;
+wmma::fragment<wmma::accumulator, 16, 16, 16, float> c_frag;
+wmma::load_matrix_sync(a_frag, a_ptr, lda);
+wmma::mma_sync(c_frag, a_frag, b_frag, c_frag);
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/cuda-kernel/SKILL.md:37]: These codes should be removed since we only supporting wgmma (hopper sm90a and all blackwell sm100+) and umma tcgen05 (blackwell sm100), including products H100, GH100, GH200, GH202, B100, B200, GB200, dgx-spark (sm120), B300, GB300, blackwell RTX(GB202/203/205)

File: .opencode/skills/cuda-kernel/SKILL.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+### TMA (Tensor Memory Accelerator) on Hopper
+```cpp
+// Create TMA descriptor for 2D tensor
+CUtensorMap tensor_map;
+// Use tma_desc.h and tma_copy_impl.h from this repo
+// See jit_kernel/csrc/thunder_moun/arch/tma/
+```
+
+## Bank Conflict Avoidance
+- Pad shared memory arrays: `__shared__ float smem[BLOCK_SIZE][BLOCK_SIZE + 1]`
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/cuda-kernel/SKILL.md:49]: Should not use this :

```
__global__ kernel_entry(...) {

...
if (condition_a) {
  kernel<...,2>(...);
} else if (condition_b){
  kernel<...,4>(...);
}
...

}
```

In this codes, when nvcc compiles kernel_entry (called in host, execute on chip), it assumed multiple device functions called inside, hence multiple share memory copies will be reserved.

Hence we should avoided this specification in a complex kernel generation.

Please follow the original design.

File: .opencode/skills/cuda-kernel/SKILL.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+- Use `float4` / `half2` for vectorized access
+- Swizzle patterns for multi-dimensional access
+
+## Warp-Level Reduction
+```cpp
+// Butterfly shuffle reduction
+for (int offset = 16; offset > 0; offset >>= 1) {
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/cuda-kernel/SKILL.md:56]: offset=WARP_SIZE/2

File: .opencode/skills/metal-kernel/SKILL.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+// SIMD group matrix multiply-accumulate (like tensor cores)
+simdgroup_float8x8 c;
+simdgroup_multiply_accumulate(c, a, b, c);
+
+// SIMD shuffle (like __shfl_xor_sync)
+float val = simd_shuffle(val, lane);
+```
+
+## 1-Bit Quantized GEMM Pattern (from PR #609)
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/metal-kernel/SKILL.md:72]: what is the PR#609 ?

File: .opencode/skills/cuda-kernel/SKILL.md (hunk 1 above)
Comment (yiakwy-xpu-ml-framework-team) [reply .opencode/skills/cuda-kernel/SKILL.md:25]: In Dgx spark (sm120) we are supporting mma with e2m1 extension.

It is not sm 80 mma, perhaps sm91 mma with wgmma tile-level shape ?

File: .opencode/skills/metal-kernel/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    N, K = weights.shape
+    pack_size = 16  # uint16 = 16 bits
+    binary = ((weights + 1) // 2).to(torch.uint16)
+    packed = torch.zeros((N, K // pack_size), dtype=torch.uint16)
+    for i in range(pack_size):
+        packed |= binary.reshape(N, -1, pack_size)[:, :, i] << i
+    return packed
+```
+
+### Bit Unpacking in Metal (XOR trick)
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/metal-kernel/SKILL.md:86]: Yes this is skills we used to develop MLX jit kernel.

File: .opencode/skills/triton-kernel/SKILL.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
+---
+name: triton-kernel
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/triton-kernel/SKILL.md:2]: Triton 3.6 is different from triton 3.5 and triton 3.4 . Hence we need to order the skills udner different trition versions.

Highly recommend skills for TVM-FFI parser.

Then gluon support.

Triton only support multi-stage kernel, while gluon (triton 3.5 and trition 3.6) support wasp kernel.

In hopper and blackwell platform, pingpong kernel is recommended in general.

File: AGENTS.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
+# Instructions for AI Agents Working on flash-float-jit-kernels
```
Comment (yiakwy-xpu-ml-framework-team) [inline AGENTS.md:1]: This file should be moved to opencode workspace, i.e. ".opencode"

File: SETUP.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
+# Setup Guide: Flash Float JIT Kernel Development Agent
```
Comment (yiakwy-xpu-ml-framework-team) [inline SETUP.md:1]: This file should be moved to opencode workspace, i.e. ".opencode"

File: .opencode/skills/cuda-kernel/SKILL.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
+---
+name: cuda-kernel
```
Comment (yiakwy-xpu-ml-framework-team) [inline .opencode/skills/cuda-kernel/SKILL.md:2]: Looking at https://github.com/sgl-project/sglang/blob/main/.claude/skills/add-jit-kernel/SKILL.md

First we need to tell opencode how to add jit kernel, then tell it the how we added and common usage.

**Review by yiakwy-xpu-ml-framework-team — COMMENTED** [review]
I will modify the codes and merge it.

File: tools/kernel_agent.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
+"""
+Simple Kernel Development Agent — CPU Edition
```
Comment (yiakwy-xpu-ml-framework-team) [inline tools/kernel_agent.py:2]: An agent should have agent loop.

Here is arch :

First you need to start  a headless agent as a proxy of LLM behind.

Kernel agent is minimum agent, but it still has agent loop.

It implements the agent loop and specifically inspect kernel files and trace remote agent (we run in port 8096 for example) to write GPU codes and execute gpu codes.

File: tools/kernel_agent.py [hunk 12]
```diff
@@ -0,0 +1,281 @@
...
+
+ROOT = pathlib.Path(__file__).parent.parent.resolve()
+sys.path.insert(0, str(ROOT))
+
+
+def check_kernel(fn, inputs) -> tuple[bool, str]:
+    """5-stage safety harness (AutoKernel paper)."""
+    import torch
+
+    # Stage 1: Smoke test
```
Comment (yiakwy-xpu-ml-framework-team) [inline tools/kernel_agent.py:28]: This implements 5-steps kernel in article . Jaber and O. Jaber, "AutoKernel: Autonomous GPU Kernel Optimization via Iterative
Agent-Driven Search," arXiv:2603.21331, 2026.

Reference must be added.

File: tools/KERNEL_AGENT.md [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+    out = torch.empty(N, dtype=x.dtype)
+    BLOCK_SIZE = 256
+    grid = (triton.cdiv(N, BLOCK_SIZE),)
+
+    from triton.runtime.interpreter import InterpretedFunction
+    InterpretedFunction(my_kernel.fn).run(...)
+    return out
+```
+
+## Safety Harness (5 Stages)
```
Comment (yiakwy-xpu-ml-framework-team) [inline tools/KERNEL_AGENT.md:61]: need references. for example,

Jaber and O. Jaber, "AutoKernel: Autonomous GPU Kernel Optimization via Iterative
Agent-Driven Search," arXiv:2603.21331, 2026.

**Review by yiakwy-xpu-ml-framework-team — CHANGES_REQUESTED** [review]

**Comment by yiakwy-xpu-ml-framework-team** [comment]
done in https://github.com/yiakwy-xpu-ml-framework-team/flash-float-jit-kernels/pull/24
